### DISCUSSION TITLE
Why are there two implementations of Maven dependency analysis?

### DISCUSSION THREAD

**Discussion by ascheman** [konveyor/analyzer-lsp#420] [General]
In the current code we find two implementations of Maven dependency analysis:
* The _normal_ [dependency analysis](https://github.com/konveyor/analyzer-lsp/blob/e73e5df290d7c74aa1e58b60c1c7356252822984/provider/internal/java/dependency.go#L46) scans the [output of an execution](https://github.com/konveyor/analyzer-lsp/blob/e73e5df290d7c74aa1e58b60c1c7356252822984/provider/internal/java/dependency.go#L58) of `mvn dependency:tree`,
* if this fails (either [by error](https://github.com/konveyor/analyzer-lsp/blob/e73e5df290d7c74aa1e58b60c1c7356252822984/provider/internal/java/dependency.go#L60), or [empty result](https://github.com/konveyor/analyzer-lsp/blob/e73e5df290d7c74aa1e58b60c1c7356252822984/provider/internal/java/dependency.go#L63)), an [internal implementation](https://github.com/konveyor/analyzer-lsp/blob/e73e5df290d7c74aa1e58b60c1c7356252822984/provider/internal/java/dependency.go#L100) analyses the `pom.xml` files directly (which is also not error-free).

My personal opinion (for discussion) is: There should be one and only one implementation for the analysis. If it fails for whatever reason this should be treated as a bug, and the bug should be fixed. Maintaining two implementations is to much overhead and confuses contributors. 

Confusion can be seen by errors like #390, which only can happen in the second case, and #392, which happen by incomplete implementation of Maven output parsing (among other problems like #393). I would even ask what makes the first one the default implemenation and the second operation the fallback? Why isn't it the other way round? Why aren't both combined to a working solution?

**Comment by ascheman** [comment]
How a root problem in the analysis leads to second problem is documented [here](https://github.com/konveyor/analyzer-lsp/issues/392#issuecomment-1808896110).

**Comment by jmle** [comment]
We might not need the fallback anymore, but I am not sure about why it was added in the first place. @pranavgaikwad do we still need it? I don't think we will, especially with the new plugin
